kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/DataEncryptionKey.java

```diff
@@ -0,0 +1,333 @@
+            this.paramSupplier = cipherSpec.paramSupplier();
+            this.cipherSpec = cipherSpec;
+            this.key = key;
+            this.numEncryptions = numEncryptions;
+        }
+
+        /**
+         * @return The encrypted key
+         */
+        public E edek() {
+            return edek;
+        }
+
+        /**
+         * Perform an encryption operation using the DEK.
+         * @param plaintext The plaintext to be encrypted
+         * @param aad The AAD to be included in the encryption
+         * @param paramAllocator A function that will return a buffer into which the cipher parameters will be written.
+         * The function's argument is the number of bytes required for the cipher parameters.
+         * @param ciphertextAllocator A function that will return a buffer into which the ciphertext will be written.
+         * The function's first argument is the number of bytes required for the cipher parameters.
+         * The function's second argument is the number of bytes required for the ciphertext.
+         * @throws DekUsageException If this Encryptor has run out of operations.
+         */
+        public void encrypt(@NonNull ByteBuffer plaintext,
+                            @Nullable ByteBuffer aad,
+                            @NonNull IntFunction<ByteBuffer> paramAllocator,
+                            @NonNull BiFunction<Integer, Integer, ByteBuffer> ciphertextAllocator) {
+
+            if (--numEncryptions >= 0) {
+                try {
+                    AlgorithmParameterSpec params = paramSupplier.get();
+                    cipher.init(Cipher.ENCRYPT_MODE, key, params);
+
+                    int size = cipherSpec.size(params);
+                    var parametersBuffer = paramAllocator.apply(size);
+                    cipherSpec.writeParameters(parametersBuffer, params);
+                    parametersBuffer.flip();
+
+                    int outSize = cipher.getOutputSize(plaintext.remaining());
+                    var ciphertext = ciphertextAllocator.apply(size, outSize);
+                    if (aad != null) {
+                        cipher.updateAAD(aad);
+                        aad.rewind();
+                    }
+                    var p = plaintext.position();
+                    cipher.doFinal(plaintext, ciphertext);
+                    plaintext.position(p);
+                    ciphertext.flip();
+
+                    if (numEncryptions == 0) {
+                        close();
+                    }
+                    return;
+                }
+                catch (GeneralSecurityException e) {
+                    throw new EncryptionException(e);
+                }
+            }
+            throw new DekUsageException("The Encryptor has no more operations allowed");
+        }
+
+        @Override
+        public void close() {
+            if (key != null) {
+                key = null;
+                maybeDestroyKey(DataEncryptionKey::releaseCryptor);
```

Comment:
The key is destroyed when we first hit a zero reference count _after `destroy()` was first called_. 
We don't destroy the key when the number of encryptions hits zero (without a call to `destroy()`), because decryptors are still allowed to use it. This allows a `DataEncryptionKey` to be shared between the caches on both the encryption and decryption path, then run out of encryptions, but remain functional in the decryption cache. Though now I write this down I think, "but who calls destroy and when?" and wonder if we need separate `closeForEncrypt()` and `closeForDecrypt()`, which we use on eviction from each cache?

Reply:
doh, end is -1 i see

Reply:
@robobario what do you make of the `closeForEncrypt()` and `closeForDecrypt()` to make it more compatible with our two caches?

Reply:
Maybe we wouldn't need to differentiate if we made the client-exposed interface:

```
CryptorManager<K, E> {
    CompletionStage<Encryptor> getEncryptor(K kek, CipherSpec cipher, int numEncrypts);
    CompletionStage<Decryptor> getDecryptor(E edek, CipherSpec cipher);
}
```

We would only hand out usable Encryptors, we would close it for encryption when a client requests numEncrypts > remaining, and obtain a new encryption DEK. Meanwhile the old DEK would still be available for decrypt.

CryptorManager would be responsible for removing idle DEKs.

Reply:
to catch our talk, to your mind we would have two caches:

1. the encryption cache, `kek -> dek`
2. the decryption cache, `edek -> dek`

on encrypt, when we obtain a new DEK we would populate both caches on the assumption that decryption is likely to occur soon (probably on another channel)

on decrypt, if we have to look up the DEK we only populate the decryption cache

these caches could have an expire after access policy and once a DEK is cleaned out of both caches we can call destroy and rely on the reference counting to do the final destroy once all it's associated cryptors have been released.

So you're suggesting to do this the reference counter needs to track both states so we can safely destroy it when it's no longer in either cache. As suggested perhaps using an integer/long to track both states and the reference count, since we are currently using the sign to track destruction.

Reply:
@robobario yes, that's a good summary. 
> 1. the encryption cache, `(kek, cipherSpec) -> dek`
> 2. the decryption cache, `(edek, cipherSpec) -> dek`

I'd overlooked that the cipherSpec is part of the cache key (damn, yet another overloading of the word key).

I think for the decryption cache we don't need the `cipherSpec` to be part of the cache key though. We encounter an `edek` by looking at a `Record`, and that `Record` will contain the `persistentId` of the cipher spec. Moreover the DekManager (in tandem with the Kms contract that generateDek return a _new_ key) already guarantees that a given DEK can only be used with a single cipherSpec. Thus it's enough for the decryption cache to be keyed on edek alone. I don't think this makes a significant difference to this PR though.

> on decrypt, if we have to look up the DEK we only populate the decryption cache

Yes, and when we do that through the `DekManager` we get back a `DataEncryptionKey` that's only good for decryption (it has 0 remaining encryptions). Which is fine, because those `DataEncryptionKey` instances never get added to the encryptionDekCache. 

> So you're suggesting to do this the reference counter needs to track both states so we can safely destroy it when it's no longer in either cache

Yes. You could argue that it's a bit of a layering violation: We're designing the DekManager around the fact that we have these two caches and that we're going to couple cache eviction to key destruction. I'm sure we could _manage_ with a single reference count in the `DataEncryptionKey`, and figure out when to call `destroy` in the place where we're doing the caching. But I strongly suspect that it would be more complex to do it there.

Reply:
I split up the reference counts and did the `closeForEncrypt()` and `closeForDecrypt()` change too.